Helpdocument<-"
#this script is to score given seq using the newly invented integration method (please see the last part of 自分で情報を集める５.docx) 
#
#Usage: R --slave --vanilla --args  -mutfile (chrpos format mut file with transcript ID) -output (output file name) < Regress_Score.v0.5.R
#
#ex) R --slave --vanilla --args --mutfile temp.test.chrpos.txt -output temp.test.chrpos.output.txt <Regress_Score.v0.5.R
#ex) R --slave --vanilla --args -seq testseq.txt -output temp.test.chrpos.output.v0.6.txt -debug <Regress_Score.v0.7.R
#ex) R --slave --vanilla --args --mutfile LMNAmut_for_gBlcok.txt -output LMNAmut_for_gBlcok.txt.regres_score.v0.7.txt -esbTranscriptID ENST00000368300 <Regress_Score.v0.7.R
#ex) R --slave --vanilla --args -canonicalSS --mutfile LMNAmut_for_gBlcok.txt -output LMNAmut_for_gBlcok.txt.regres_score.v0.8.txt -esbTranscriptID ENST00000368300 <Regress_Score.v0.8.R
#ex) R --slave --vanilla --args -canonicalSS -seq LMNAc768GAseq.txt -output LMNAc768GA.regres_score.v0.8.txt -debug <Regress_Score.v0.8.R
#ex) R --slave --vanilla --args --mutfile JamesTTNvariant.txt -output JamesTTNvariant.regres_score.v0.83.txt -entiregene -ss3 -debug < Regress_Score.v0.83.R
#ex) R --slave --vanilla --args --mutfile JamesTTNvariant.txt -output JamesTTNvariant.3000.regres_score.v0.84.txt -wide 3000 -ss3 -debug < Regress_Score.v0.84.R
#ex) R --slave --vanilla --args --mutfile JamesTTNvariant.txt -output JamesTTNvariant.wide1000.regres_score.v0.84.txt -wide 1000 -ss3 -debug < Regress_Score.v0.84.R
#ex) R --slave --vanilla --args --mutfile JamesTTNvariant.txt -output JamesTTNvariant.regres_score.v0.84.txt -entiregene -ss3 -debug -wide 1000 < Regress_Score.v0.84.R
#ex) R --slave --vanilla --args -mutfile test.input1.txt -output test.output.v0.87.txt < Regress_Score.v0.87.R
#ex) R --slave --vanilla --args -mutfile ./Exac/test.txt -output ./Exac/Exac.r0.1-0.3.TTN.SNV.txt.Regress_Score.v0.87.txt -esbTranscriptID ENST00000342992 -lseq ./Exac/ENST00000342992.txt -ss5 -skipRegressScore -skipSRE< Regress_Score.v0.88.R
#Options:
#-seq (a file contains sequence data)  exon should be UPPERCASE, intron should be LOWERCASE, you can indicate mutation like (A|B), where A means refseq and B mean alt seq
#                                       Also you can indicate nucleutide of interest like ATGC[ATGC]ATGCa              
#-mutfile (chrpos format mut file with transcript ID) e.g.chr1:123G>A ENST000001234 in v0.5 just 1 line accept
#-output (filename to be saved as a result)
#-marthost (indicates biomaRt server)
#-esbTranscriptID (transcriptID)  when you check cDNA variant list, you should provide transcriptID
#-nomart        if you indicate this switch, connecting biomaRt is being skipped.
#-canonicalSS  indicate this script to process canonical 5ss and 3ss around the given mutation, if the mutation is not on the exon, this function does not work.
#-entiregene indicates this script wull calculate entire gene score 
#-ss5 indicates this script to calculate ss5 scores
#-ss3 indicates this script to calculate ss3 scores
#     if omitted the both above automatically calculate both
#-wide (integer) indicates to make scan region added (integer) around the mutation
#-lseq (filename) indicagtes transcript info generated by GetSeq(TranscriptID,1) if you employ this, you don't need to connect biomart server , which make run time short.
#                 whme you use it, I recommned you to use -esbTranscriptID to indicate the name of transcript.
#-skipRegressScore indicates skipping regress score calculation to make time short. The score will be zero
#-skipSRE indicates skipping checking # of splicing regulatory elements. The # of elements will be zero
#most of element search sub routines came from DataCollecterPartA.v0.9.R and they had been modified to fit this current situation.
#most of constructing ref seq and searching alt seq code came from 5ss_cal_ver1.19.R
#
#History
#v0.5 launch ver.
#v0.6 can process submitted sequence.   can process nearby authentic SS.
#v0.7 can process both chr pos and cDNA format
#v0.8 can indicate to process canonical splice site using -canonicalSS
#v0.81 can indicate to process entire gene using -entiregene
#v0.82 can indicate either 5ss or 3ss calculation
#v0.83 bug-fix for reverse strand gene (fix the process going up accoding to phypos-> according to cDNApos)
#v0.84 -wide check added
#v0.85 when transcripID NOT given in chrpos mode, automatically find all transcritps for it and try all of them
#         (change sub-routine findtranscript and chrposmut)
#     skip variants NOT on the transcript e.g. downstream varint or UTRs...
#     bug-fix
#v0.86 check increase or decrease # of ESE,ESS, Grun and ISE 250bp around the mut for comprehensive analysis. just columns added in the result. So you can employ old-scripts for v0.84 or 0.85
#v0.87 bug-fix  v0.86 counts # of splicing elements on the putative exon-intron generated by putative splicing variants. In this ver, just count # of elements around the variant
#       #ESE,ESS derived from exon where the mutation is.      #Grun,ISE derived from intron where the mutation is.  e.g.)if the mutation is on exon, #Grun, #ISE must be 0.
# Accept ENSEMBL transcriptID 9e.g.ENST0000001234 and RefseqmRNAID e.g.NM_001195597  (if NM_001195597.2 given, automatically take just NM_001195597 to avoid error)
#     when 1 NM ID corresponds to multiple ENST ID, just take the 1st one (->GetSeq)
#v0.88 stand alone mode added, which means when you process one transcript and you have seq file for it, just indicate that and read lseq info from file and not connect to biomaRt
#     it can speed up process and run more nodes
"

#Subs
inputparameters<-function(inputlist,inputparas) {  #input parameters from commnand line
  if (length(inputlist) != length(inputparas)) stop("The length NOT match in the input parameters\n")
  inputparasrev<-rep(-9,length(inputparas))  #check value if the value is already input or not
  for (i in 1:length(commandArgs(trailingOnly=TRUE))) { 
    for (j in 1:length(inputlist)) {
      if (commandArgs(trailingOnly=TRUE)[i] == inputlist[[j]] & inputparasrev[j]==-9) {
        inputparasrev[j]<-1
        if (inputparas[j]==0) {  #input sw 1
          inputlist[[j]]<-1 
        } else { #input a series of value or character
          inputlist[[j]]<-commandArgs(trailingOnly=TRUE)[i+1]
        }
      }
    }    
  }
  
  for (i in 1:length(inputparasrev)) {  #when no input, then assign -9 for null parameter (-9 is a null value)
    if (inputparasrev[i]==-9) inputlist[[i]]<-as.numeric(-9)
  }
  return (inputlist)
}

positionSeq<-function(Seq,start,end, oristart,oriend,strand) {  # this script is to figure out where actual start and end position in the given sequence and return corresponding region indicated by start-end   #given that seq spans oristart-oriend, calculate where start and end locates on the seq.
  if (start>0 & end>0) {
    if (strand==1) { # forward strand
      acstart<-start-oristart+1
      acend<-end-oristart+1
    } else { #reverse strand
      acstart<-oriend-end+1
      acend<-oriend-start+1
    }
    return(substr(Seq,acstart,acend))
  } else {
    return("")
  }
}

#sub routine for 5ss and 3ss maxent
makemaxentscores3<-function(){
  dir<-"./ssfiles"
  list<-c('me2x3acc1','me2x3acc2','me2x3acc3','me2x3acc4','me2x3acc5','me2x3acc6','me2x3acc7','me2x3acc8','me2x3acc9')
  metables<-list()
  num<-0
  initial<-1
  for (i in list) {
    vec<-c()
    SCOREF<-readLines(con=paste(dir,i,sep="/"))
    for (j in 1:length(SCOREF)) {
      vec<-c(vec,as.numeric(SCOREF[j]))
    }
    new<-list(val=I(vec))
    names(new)<-num
    if (initial==1) {metables<-new;initial<-0;} else metables<-c(metables,new)  # row correspond to n and col correspond to num
    num<-num+1
  }
  return (metables)
}

makemaxentscores5<-function(){
  me2x5<-read.table("./ssfiles/me2x5",header=F,stringsAsFactors=F)  #if you want to make the running time short, you should make this variable in the first of your script
  seq<-read.table("./ssfiles/splice5sequences",header=F,stringsAsFactors=F)
  scorematrix<-cbind(seq,me2x5)
  names(scorematrix)<-c("Seq","Score")
  return(scorematrix)
}

hashseq<-function(seq){
  seq<-toupper(seq)
  seq<-gsub("A","0",seq)
  seq<-gsub("C","1",seq)
  seq<-gsub("G","2",seq)
  seq<-gsub("T","3",seq)
  seqa<-unlist(strsplit(seq,""))
  sum<-1
  len<-nchar(seq)
  four<-c(1,4,16,64,256,1024,4096,16384)
  i<-0
  while (i<len) {
    sum<-sum+as.numeric(seqa[i+1])*four[len-i]
    i<-i+1
  }
  return(sum)
}

getrest<-function(seq) {
  return(paste(substr(seq,1,18),substr(seq,21,23),sep=""))
}

scoreconsensus3sub<-function(seq) {
  seqa<-unlist(strsplit(seq,""))
  bgd<-list(A=0.27,C=0.23,G=0.23,T=0.27)
  cons1<-list(A=0.9903,C=0.0032,G=0.0034,T=0.0030)
  cons2<-list(A=0.0027,C=0.0037,G=0.9905,T=0.0030)
  addscore<-as.numeric(cons1[seqa[19]])*as.numeric(cons2[seqa[20]])/(as.numeric(bgd[seqa[19]])*as.numeric(bgd[seqa[20]]))
  return(addscore)
}

scoreconsensus5sub<-function(inputseq) {
  sc34<-data.frame(ATGC=c("A","C","G","T"),
                   bgd=c(0.27,0.23,0.23,0.27),
                   cons1=c(0.004,0.0032,0.9896,0.0032),
                   cons2=c(0.0034,0.0039,0.0042,0.9884))
  chr3<-substr(inputseq,4,4)
  chr4<-substr(inputseq,5,5)
  score<-sc34[sc34$ATGC==chr3,]$cons1*sc34[sc34$ATGC==chr4,]$cons2/(sc34[sc34$ATGC==chr3,]$bgd*sc34[sc34$ATGC==chr4,]$bgd); 
  return(score)
}

maxentscore3sub<-function(seq,metables) {
  sc0<-as.numeric(unlist(metables[1])[hashseq(substr(seq,1,7))])
  sc1<-as.numeric(unlist(metables[2])[hashseq(substr(seq,8,14))])
  sc2<-as.numeric(unlist(metables[3])[hashseq(substr(seq,15,21))])
  sc3<-as.numeric(unlist(metables[4])[hashseq(substr(seq,5,11))])
  sc4<-as.numeric(unlist(metables[5])[hashseq(substr(seq,12,18))])
  sc5<-as.numeric(unlist(metables[6])[hashseq(substr(seq,5,7))])
  sc6<-as.numeric(unlist(metables[7])[hashseq(substr(seq,8,11))])
  sc7<-as.numeric(unlist(metables[8])[hashseq(substr(seq,12,14))]) 
  sc8<-as.numeric(unlist(metables[9])[hashseq(substr(seq,15,18))]) 
  
  finalscore<-sc0*sc1*sc2*sc3*sc4/(sc5*sc6*sc7*sc8)
  return(finalscore)
}

#main function of maxent
maxent3score<-function(seq) {
  return(log2(scoreconsensus3sub(seq)*maxentscore3sub(getrest(seq),metables)))
}

maxent5score<-function (seq) {
  getrest<-paste(substr(seq,1,3),substr(seq,6,9),sep="")
  return(log2(scoreconsensus5sub(seq)*as.numeric(scorematrix[scorematrix$Seq==getrest,]$Score)))
}

#checking accessory introns e.g. G run, ISE
intronscan<-function(seq,beforeafter,skip) {   #seq -intronseq ,beforeafter  the intron exists before a exon of interest (-1) or after a exon of interest (1) ,skip 20(5 end of exon) or 6 (3 end of exon)
  output<-list(Grun=0,ISE=0)  #reply scores of intron elements
  gflag<-0
  scanstart<-1+(beforeafter==1)*skip
  scanend<-nchar(seq)-(beforeafter==-1)*skip
  for (i in scanstart:scanend) {
    basepos<-ifelse(beforeafter==1,i,nchar(seq)-i+1)
    if (substr(seq,i,i)=="G") { #checking G run
      gflag<-gflag+1
      if (i==nchar(seq) & gflag>2) {
        output$n_Grun<-output$n_Grun+1
        pos<-basepos-gflag*(beforeafter==1)  #pos means the distance from the exon
        output$Grun<-output$Grun+PosScore(pos,ifelse(beforeafter==1,Eweight$aGrun,Eweight$bGrun))
      }
    } else if (gflag<3) {
      gflag<-0
    } else {
      output$n_Grun<-output$n_Grun+1
      pos<-basepos-gflag*(beforeafter==1)  #pos means the distance from the exon
      output$Grun<-output$Grun+PosScore(pos,ifelse(beforeafter==1,Eweight$aGrun,Eweight$bGrun))
      gflag<-0
    }
    tempISE9<-ifelse ((i+9-1)<=scanend,substr(seq,i,i+9-1),"dummy") #checking ISE9 seq
    if (length(ISE9seq[ISE9seq==tempISE9])>0) {
      output$n_ISE<-output$n_ISE+1
      pos<-basepos-8*(beforeafter==-1) #pos means the distance from the exon
      output$ISE<-output$ISE+PosScore(pos,ifelse(beforeafter==1,Eweight$aISE,Eweight$bISE))
     }
    tempISE10<-ifelse ((i+10-1)<=scanend,substr(seq,i,i+10-1),"dummy") #checking ISE10 seq
    if (length(ISE10seq[ISE10seq==tempISE10])>0) {
      output$n_ISE<-output$n_ISE+1
      pos<-basepos-9*(beforeafter==-1) #pos means the distance from the exon
      output$ISE<-output$ISE+PosScore(pos,ifelse(beforeafter==1,Eweight$aISE,Eweight$bISE))
    }    
  }
  return(output)
}

#exonscan 3ss-seq & 5ss-seq, ESE and ESS on the exon, G run on both introns surrounding the exon
exonscan<-function(seq,from) {  #Seq a sequence from the edge of exon,  from 5 or 3
  maxlen<-nchar(seq)
  exonlen<-maxlen
  output<-list(ESE=0,ESS=0)  #reply scores of exon elements
  for (j in 1:(maxlen-min(ESElength,ESSlength)))  {#checking ESE &ESS seqs on the exon
    
    tempESE<-ifelse ((j+ESElength-1)<=maxlen,substr(seq,j,j+ESElength-1),"dummy")
    tempESS<-ifelse ((j+ESSlength-1)<=maxlen, substr(seq,j,j+ESSlength-1),"dummy")
    #ESpos<-ifelse(Strand==1,start+j-1,end-j+1)  #ES position is given by phypos
    ESpos<-j  #ES position is given by the distance from 5'start of the exon  start pos is 1
    if (length(ESEseq[ESEseq==tempESE])>0) {
      ESposR<-exonlen-ESpos-ESElength+1 #ESposR means a distance from 3' end of the exon
      pos<-ifelse(from==5,ESpos,ESposR)
      output$ESE<-output$ESE+PosScore(pos,ifelse(from==5,Eweight$ESE5,Eweight$ESE3))
    }
    if (length(ESSseq[ESSseq==tempESS])>0) {
      ESposR<-exonlen-ESpos-ESSlength+1 #ESposR means a distance from 3' end of the exon
      pos<-ifelse(from==5,ESpos,ESposR)
      output$ESS<-output$ESS+PosScore(pos,ifelse(from==5,Eweight$ESS5,Eweight$ESS3))
    }
  }
  return(output)
}

PosScore<-function(pos,ScoreVec) {  #calculate a score for each element, ScoreVec is "Mean,SD" (should be given as charcter) for gaussian model.  multi Mean,Sds are acceptable e.g. mean,sd,mean,sd,....
  meansd<-as.numeric(unlist(strsplit(ScoreVec,",")))
  score<-sum(dnorm(pos,mean=meansd[seq(1,length(meansd),by=2)],sd=meansd[seq(2,length(meansd),by=2)]))
  return(score)
}

SSscan<-function(seq,ExPosInTheContext,SS)  {#seq a sequence including 5ss or 3ss, ExPosInTheContext exon position(Start,End) in the given seq  SS should be 5(5Splice site) or 3 (3Spice site) 
  startend<-as.numeric(unlist(strsplit(ExPosInTheContext,",")))
  seq<-toupper(seq)
  output<-list(Maxent=0,SSscore=0,SSseq="")
  if (SS==5) { #5ss processing
    ss5seq<-substr(seq,startend[2]-2,startend[2]+6)
    if (nchar(ss5seq)==9) output$Maxent<-maxent5score(ss5seq) else output$Maxent<--99
    #cat("Submit exon:",substr(seq,startend[1],startend[2]), " Submit intron:",substr(seq,startend[2]+1,nchar(seq)),sep="")
    output$SSscore<--99
    if (IN$skipRegressScore==-9) {
      Escore<-exonscan(substr(seq,startend[1],startend[2]),3)
      Iscore<-intronscan(substr(seq,startend[2]+1,nchar(seq)),1,6)
      logOdds<-RegCoef5$Intercept+RegCoef5$MAXENT5*output$Maxent+RegCoef5$ESE3*Escore$ESE+RegCoef5$ESS3*Escore$ESS+RegCoef5$aGrun*Iscore$Grun+RegCoef5$aISE*Iscore$ISE
      if (output$Maxent!=-99) output$SSscore<-1/(1+exp(-logOdds))
    }
    output$SSseq<-paste(substr(ss5seq,1,3),tolower(substr(ss5seq,4,9)),sep="")
    #cat("SS5seq:",substr(ss5seq,1,3),tolower(substr(ss5seq,4,9)),"\n",sep="")
  } else { #3ss processing
    ss3seq<-substr(seq,startend[1]-20,startend[1]+2)
    #cat("SS3seq:",ss3seq,"\n")
    if (nchar(ss3seq)==23) output$Maxent<-maxent3score(ss3seq) else output$Maxent<--99
    #cat("Submit exon:",substr(seq,startend[1],startend[2]), " Submit intron:",substr(seq,1,startend[1]-1),sep="")
    output$SSscore<--99
    if (IN$skipRegressScore==-9) {
      Escore<-exonscan(substr(seq,startend[1],startend[2]),5)
      Iscore<-intronscan(substr(seq,1,startend[1]-1),-1,20)
      logOdds<-RegCoef3$Intercept+RegCoef3$MAXENT3*output$Maxent+RegCoef3$ESE5*Escore$ESE+RegCoef3$ESS5*Escore$ESS+RegCoef3$bGrun*Iscore$Grun+RegCoef3$bISE*Iscore$ISE
      if (output$Maxent!=-99) output$SSscore<-1/(1+exp(-logOdds)) 
    }
    output$SSseq<-paste(tolower(substr(ss3seq,1,20)),substr(ss3seq,21,23),sep="")
    #cat("SS3seq:",tolower(substr(ss3seq,1,20)),substr(ss3seq,21,23),"\n",sep="")
  } 
  return(output)
}

SEEscan<-function(seq) {
  output<-list(nESE=0,nESS=0)
  maxlen<-nchar(seq)
  for (j in 1:(maxlen-min(ESElength,ESSlength)))  {#checking ESE &ESS seqs on the exon
    tempESE<-ifelse ((j+ESElength-1)<=maxlen,substr(seq,j,j+ESElength-1),"dummy")
    tempESS<-ifelse ((j+ESSlength-1)<=maxlen, substr(seq,j,j+ESSlength-1),"dummy")
    #ESpos<-ifelse(Strand==1,start+j-1,end-j+1)  #ES position is given by phypos
    ESpos<-j  #ES position is given by the distance from 5'start of the exon  start pos is 1
    if (length(ESEseq[ESEseq==tempESE])>0) output$nESE<-output$nESE+1
    if (length(ESSseq[ESSseq==tempESS])>0) output$nESS<-output$nESS+1
  }
  return(output)
}

SEIscan<-function(seq) {
  output<-list(nGrun=0,nISE=0)
  gflag<-0
  for (i in 1:nchar(seq)) {
    if (substr(seq,i,i)=="G") { #checking G run
      gflag<-gflag+1
      if (i==nchar(seq) & gflag>2) {
        output$nGrun<-output$nGrun+1
      }
    } else if (gflag<3) {
      gflag<-0
    } else {
      output$nGrun<-output$nGrun+1
      gflag<-0
    }
    tempISE9<-ifelse ((i+9-1)<=nchar(seq),substr(seq,i,i+9-1),"dummy") #checking ISE9 seq
    if (length(ISE9seq[ISE9seq==tempISE9])>0) {
      output$nISE<-output$nISE+1
    }
    tempISE10<-ifelse ((i+10-1)<=nchar(seq),substr(seq,i,i+10-1),"dummy") #checking ISE10 seq
    if (length(ISE10seq[ISE10seq==tempISE10])>0) {
      output$nISE<-output$nISE+1
    }    
  }
  return(output)
}

SEscan<-function(newseq,cDNApos) {
  exonseq<-"";
  intronseq<-"";
  #name intron as -1 , -2 ,-3...
  if (cDNApos[1]<0 | IN$skipSRE==1) {
    return(list(nESE=-99,nESS=-99,nGrun=-99,nISE=-99))
  } else {
    #name intron to identify each intron separately
    exons<-unique(newseq$exon)
    exons<-sort(exons[exons>0])
    for (i in exons) {
      minpos<-min(which(newseq$exon==i))
      maxpos<-max(which(newseq$exon==i))
      if (i<max(exons)) {
        nextmin<-min(which(newseq$exon==i+1))
        newseq$exon[(maxpos+1):(nextmin-1)]=-i
      }
      if (i==1) {
        if (minpos>1) newseq$exon[1:(minpos-1)]=0
      } 
    }
  takeregion<-unique(newseq$exon[cDNApos[1]:cDNApos[2]])
  takeexon<-takeregion[takeregion>0]
  takeintron<-takeregion[takeregion<1]
  SEE<-list(nESE=0,nESS=0)
  if (length(takeexon)>0) {
    seq<-""
    for (i in takeexon) {
      seq<-paste(seq,paste(newseq$Seq[newseq$exon==i],collapse=""),"XXXXXXXXXXXX")
    }
    SEE<-SEEscan(seq)
  }
  SEI<-list(nGrun=0,nISE=0)
  if (length(takeintron)>0) {
    seq<-""
    for (i in takeintron) {
      seq<-paste(seq,paste(newseq$Seq[newseq$exon==i],collapse=""),"XXXXXXXXXXXX")
    }
    SEI<-SEIscan(seq)
  }
  return(list(nESE=SEE$nESE,nESS=SEE$nESS,nGrun=SEI$nGrun,nISE=SEI$nISE))
  }
}

SS5scan<-function(newseq,Start,End,cDNApos) {
  #cat("Seq: ",tolower(paste(newseq$Seq[1:(Start-1)],collapse="")),paste(newseq$Seq[Start:End],collapse=""),tolower(paste(newseq$Seq[(End+1):nrow(newseq)],collapse="")), " Start:",Start," End:",End,"\n")
  result5<-data.frame()
  if (Start>End) {temp<-c(Start,End);Start<-temp[2];End<-temp[1]}
  for (j in Start:End) {
    exonend<-j+2    #putative exon end generated by the mutation
    if (exonend>=3 & exonend<=nrow(newseq)) { 
    exonrank<-newseq$exon[exonend] # #exon where the above exon end exists
    if (exonrank==-9 ) exonrank<-max(newseq$exon[1:exonend])
    if (exonrank>0) {
      exonstart<-min(which(newseq$exon==exonrank))   #  exon start pos in the context of seq
    } else {
      exonstart<-1
    }
    a<-which(newseq$exon==(exonrank+1));if (length(a)>0) intronend<-min(a)-1 else  intronend<-nrow(newseq)
    ###landscape is   seq(exonstart....exonend....intronend)
    exonstart_posincontext<-1
    if ((exonend-exonstart+1)<250) {
      takestart<-exonstart;exonend_posincontext<-exonend-exonstart+1
      #cat("5within\n")
    } else {
      takestart<-exonend-250;exonend_posincontext<-250+1
    }
    if ((intronend-exonend)<250) {
      takeend<-intronend
      #cat("3within\n")
    } else {
      takeend<-exonend+250
    }
    seq<-paste(newseq$Seq[takestart:takeend],collapse ="")
    #cat("j;",j,"takestart",takestart,"takeend ", takeend,"\n")
    #cat("Seq: ",tolower(substr(seq,1,exonstart_posincontext-1)),substr(seq,exonstart_posincontext,exonend_posincontext),tolower(substr(seq,exonend_posincontext,nchar(seq))),"\n",sep="")
    output<-SSscan(seq,paste(exonstart_posincontext,exonend_posincontext,sep=","),5)
    cDNAboundary<-paste("c.",newseq$cDNApos[exonend],"-",newseq$cDNApos[exonend+1],sep="")
    phyposboundary<-paste("chr:",newseq$chr[exonend],":",newseq$phypos[exonend],"-",newseq$phypos[exonend+1],sep="")
    output2<-SEscan(newseq,cDNApos)
    #cat("len newseqphypos",length(newseq$phypos[j])," nrow output:",unlist(output), " length cdnaboundary:",length(cDNAboundary)," length phyposboundary:",length(phyposboundary),"\n")
    new<-data.frame(phypos=newseq$phypos[j],MaxEntScoreSS5=output$Maxent,SS5score=output$SSscore,SS5cDNAboundary=cDNAboundary,SS5phyposboundary=phyposboundary,SS5seq=output$SSseq,ESE5=output2$nESE,ESS5=output2$nESS,Grun5=output2$nGrun,ISE5=output2$nISE)
    result5<-rbind(result5,new)
    }
  }
  return(result5)
}

SS3scan<-function(newseq,Start,End,cDNApos) {
  #cat("Seq: ",tolower(paste(newseq$Seq[1:(Start-1)],collapse="")),paste(newseq$Seq[Start:End],collapse=""),tolower(paste(newseq$Seq[(End+1):nrow(newseq)],collapse="")), " Start:",Start," End:",End,"\n")
  result3<-data.frame()
  if (Start>End) {temp<-c(Start,End);Start<-temp[2];End<-temp[1]}
  for (j in Start:End) {
    exonstart<-j+20    #putative exon end generated by the mutation
    if (exonstart>=21 & exonstart<=nrow(newseq)) { 
      #cat("exonstart:",exonstart,"\n")
    exonrank<-newseq$exon[exonstart] # #exon where the above exon end exists
    if (exonrank==-9) {
      temp<-sort(unique(newseq$exon[exonstart:nrow(newseq)]))
      if (length(temp)>1) exonrank<-temp[2] 
    }  
    if (exonrank>0) {
      exonend<-max(which(newseq$exon==exonrank))   #  exon end pos in the context of seq
    } else {
      exonend<-nrow(newseq)
    }
    a<-which(newseq$exon==(exonrank-1));if (length(a)>0) intronstart<-max(a)+1 else  intronstart<-1
    ###landscape is   seq(intronstart...exonstart...exonend)
    
    if ((exonstart-intronstart)<250) {
      takestart<-intronstart;exonstart_posincontext<-exonstart-intronstart+1
    } else {
      takestart<-exonstart-250;exonstart_posincontext<-250+1
    }
    if ((exonend-exonstart)<250) {
      takeend<-exonend;
    } else {
      takeend<-exonstart+250-1
    }
    seq<-paste(newseq$Seq[takestart:takeend],collapse ="")
    exonend_posincontext<-nchar(seq)
    #cat("Seq: ",tolower(substr(seq,1,exonstart_posincontext-1)),substr(seq,exonstart_posincontext,exonend_posincontext),tolower(substr(seq,exonend_posincontext,nchar(seq))),"\n",sep="")
    output<-SSscan(seq,paste(exonstart_posincontext,exonend_posincontext,sep=","),3)
    cDNAboundary<-paste("c.",newseq$cDNApos[exonstart-1],"-",newseq$cDNApos[exonstart],sep="")
    phyposboundary<-paste("chr:",newseq$chr[exonstart],":",newseq$phypos[exonstart-1],"-",newseq$phypos[exonstart],sep="")
    output2<-SEscan(newseq,cDNApos)
    new<-data.frame(phypos=newseq$phypos[j],MaxEntScoreSS3=output$Maxent,SS3score=output$SSscore,SS3cDNAboundary=cDNAboundary,SS3phyposboundary=phyposboundary,SS3seq=output$SSseq,ESE3=output2$nESE,ESS3=output2$nESS,Grun3=output2$nGrun,ISE3=output2$nISE)
    result3<-rbind(result3,new)
    }
  }
  return(result3)
}

GetSeq<-function(transciptID,cDNAsw) { #ID -transcriptID, Seq -each allele, phypos -physical position,cDNApos-cDNA position, exon-#exon
  if (IN$lseq==-9) {
  transcriptkind<-""
  if (length(grep("ENST",transciptID))>0) transcriptkind<-"ensembl_transcript_id"
  if (length(grep("NM_",transciptID))>0) {
    transcriptkind<-"refseq_mrna"
    if (length(grep("\\.",transciptID))>0) {
      temp<-unlist(strsplit(transciptID,"\\."))
      transciptID<-temp[1]
    }
  }
  transcriptinfo<-getBM(attributes=c("hgnc_symbol","ensembl_transcript_id","chromosome_name","transcript_start","transcript_end","strand"),
                        transcriptkind, values=transciptID, mart=mart)
  if (nrow(transcriptinfo)==0 ) return ( data.frame() )
  if(nrow(transcriptinfo)>1 ) {
    cat("Multiple transcript found ",transcriptinfo$ensembl_transcript_id,"from given ID:",transciptID, "\tThen take 1st one...\n")
    transcriptinfo<-transcriptinfo[1,]
    transcriptkind<-"ensembl_transcript_id"
    transciptID<-transcriptinfo$ensembl_transcript_id
  }
  exonsinfo<-getBM(attributes=c("ensembl_transcript_id","ensembl_exon_id","rank","chromosome_name","exon_chrom_start","exon_chrom_end"),
                   transcriptkind, values=transciptID, mart=mart)
  utr5info<-getBM(attributes=c("5_utr_start","5_utr_end"),transcriptkind, values=transciptID, mart=mart)
  utr3info<-getBM(attributes=c("3_utr_start","3_utr_end"),transcriptkind, values=transciptID, mart=mart)
  utr5info<-utr5info[!is.na(utr5info[,1]),]
  utr3info<-utr3info[!is.na(utr3info[,1]),]
  utrinfo<-data.frame(start=c(utr5info$"5_utr_start",utr3info$"3_utr_start"),end=c(utr5info$"5_utr_end",utr3info$"3_utr_end"))
  lseqtemp     <- getSequence(id = transciptID, type=transcriptkind, mart =mart, seqType = "transcript_exon_intron") 
  if (nrow(lseqtemp)>0) {
    lseqlength<-nchar(lseqtemp$transcript_exon_intron)
  } else {
    lseqlength<-0
    cat("TranscriptID:", transciptID,"was NOT found in the sequence database.\n")
  }
  if (transcriptinfo$strand==-1) {  #for reverse strand genes
    cat("Reverse strand gene found.\n")
    reverseflag<<-1 #universal variable to represent the gene is reversestrand genes
    temp<-transcriptinfo$transcript_start
    temp2<-transcriptinfo$transcript_end
    transcriptinfo$transcript_start<-temp2
    transcriptinfo$transcript_end<-temp
  } else {
    reverseflag<<-0
  }
  #cat("Lseqlength:",lseqlength,"phypos start,end:",transcriptinfo$transcript_start,",",transcriptinfo$transcript_end," transcript length by start end:",length(transcriptinfo$transcript_start:transcriptinfo$transcript_end),"\n")
  if (lseqlength==length(transcriptinfo$transcript_start:transcriptinfo$transcript_end)) {# check the consistency between annotated length (by start and end position) and actual sequence}
    lseq<-data.frame(GENE=rep(transcriptinfo$hgnc_symbol,lseqlength),ID=rep(transciptID,lseqlength),chr=rep(transcriptinfo$chromosome_name,lseqlength),phypos=transcriptinfo$transcript_start:transcriptinfo$transcript_end,cDNApos=rep(-9,lseqlength),exon=rep(-9,lseqlength),Seq=I(unlist(strsplit(lseqtemp$transcript_exon_intron,""))))
    for (i in 1:nrow(exonsinfo)) {  #mark exon number
      start<-which(lseq$phypos==exonsinfo[i,]$exon_chrom_start)
      end<-which(lseq$phypos==exonsinfo[i,]$exon_chrom_end)
      lseq[start:end,]$exon<-i
    }
    if (cDNAsw) {
      cdna.frame<-data.frame()
      for (i in 1:nrow(exonsinfo)) {  #integrate exon info and utr info then define cDNA region
        start<-exonsinfo[i,]$exon_chrom_start
        end<-exonsinfo[i,]$exon_chrom_end
        all<-0
        if (nrow(utrinfo)>0) for (j in 1:nrow(utrinfo)) {
          if (start<utrinfo[j,]$end & utrinfo[j,]$end<end) start<-utrinfo[j,]$end+1
          if (start<utrinfo[j,]$start & utrinfo[j,]$start<end) end<-utrinfo[j,]$start-1
          if (utrinfo[j,]$start<=start & end<=utrinfo[j,]$end) all<-1
        }
        if (all==0) {
          new<-data.frame(start=start,end=end,exon=i)
          #cat("New region start:",start," end:",end,"exon:",i)
          cdna.frame<-rbind(cdna.frame,new)
        }
      } 
      if (transcriptinfo$strand==-1) {  #for reverse strand genes
        cdna.frame<-data.frame(start=cdna.frame$end,end=cdna.frame$start,exon=cdna.frame$exon)
      }
      cDNApos<-1
      for (i in 1:nrow(cdna.frame)) { #actual cDNA mark
        start<-which(lseq$phypos==cdna.frame[i,]$start)
        end<-which(lseq$phypos==cdna.frame[i,]$end)
        lseq[start:end,]$cDNApos<-cDNApos:(cDNApos+abs(end-start))
        cDNApos<-cDNApos+abs(end-start)+1
        new<-data.frame(ID=transciptID,cDNApos=paste("c.",cDNApos-1,"_",cDNApos,sep=""),exon=paste("5ss_exon",cdna.frame[i,]$exon,"_",cdna.frame[i,]$exon+1))
        exon.frame<<-rbind(exon.frame,new)  #record original 5ss between exons
      } 
    }
  } else {  #if inconsistency between  between annotated length (by start and end position) and actual sequence was found 
    cat("The length based on the sequence:",lseqlength,"\t The start and end position:",transcriptinfo$transcript_start,",",transcriptinfo$transcript_end,"\t The length based on the start and end positions:",length(transcriptinfo$transcript_start:transcriptinfo$transcript_end),"\nInconsistency was found. Then an empty sequence returned to avoid further error.\n")
    lseq<-data.frame()
  }
  } else {
    lseq<-masterlseq
  }
  return(lseq)
}

findtranscript <- function (chr,pos,mutname) {
  #targetval<-paste(mut.frame[i,]$chr,":",mut.frame[i,]$pos,":",mut.frame[i,]$pos,sep="")
  targetval<-paste(chr,":",pos,":",pos,sep="")
  cat ("Searching Transcripts for",mutname, "\tSubmitting string:",targetval,"\t" )
  transcripts<-getBM(attributes=c("ensembl_transcript_id","transcript_start","transcript_end"),
                     "chromosomal_region", values=targetval, mart=mart)
  transcripts<-transcripts[transcripts$transcript_start <= pos & pos <= transcripts$transcript_end,]
  if (nrow(transcripts)>0) {
    cat("Found Transcripts:",transcripts$ensembl_transcript_id,"\n")
    return(transcripts$ensembl_transcript_id)
  } else {
    cat("No trascripts for",mutname,"\n")
    return(c())
  }
}

chrposmut<-function(muttemp){  #build a matrix for mutation e.g. chr1:11233A>G (TranscriptID)
  mut.frame<-data.frame()
  for (i in 1:length(muttemp)) {
    muttemp[i]<-gsub("^(\\s+)|(\\s+)$","",muttemp[i])
    pretemp<-unlist(strsplit(muttemp[i]," "))
    if (length(pretemp)>1) transID<-pretemp[2] else transID<-NA
    temp<-unlist(strsplit(pretemp[1],":"))
    chr<-temp[1]
    chr<-sub("chr","",chr)
    pos<-gsub("(A|T|G|C|>)","",temp[2])
    alleles <-sub("\\d+","",temp[2])
    temp3<-unlist(strsplit(alleles,">"))
    ref<-temp3[1]
    alt<-temp3[2]
    #skipflag<-0  #In ver0.9 chr1:112233_A>AAG or indels are ignored  #In ver1.12 indels in vcf file came to be accepted
    cat ("Reading Line #",i," :",muttemp[i])
    #if (nchar(ref)>1 | nchar(alt)>1) {skipflag<-1;cat("\tSkipped\n");}  
    #if (skipflag==0) {
    if (length(grep("^(A|T|G|C|.)+$",ref))>0 & length(grep("^(A|T|G|C|.)+$",alt))>0) {
      cat("\tAccepted\n")
      if (is.na(transID)) {
        if (IN$esbTranscriptID!=-9) { #when transID NOT given in Mut file and esbTranscriptID DOES given in -esbTranscriptID switch
          transID<-IN$esbTranscriptID
          cat("TranscriptID given in -esbTranscriptID:",IN$esbTranscriptID,"applied\n")
        } else {
          transID<-findtranscript(chr,pos,muttemp[i])
        }
      } 
      tlen<-length(transID)
      new.frame<-data.frame()
      if (tlen>0) new.frame<-data.frame(kind=I(rep("chrpos",tlen)),chr=I(rep(chr,tlen)),pos=I(rep(as.integer(pos),tlen)),ref=I(rep(ref,tlen)),alt=I(rep(alt,tlen)),name=I(rep(paste("chr",chr,":",pos,ref,">",alt,sep=""),tlen)),transID=I(transID))
      if (nrow(mut.frame)==0) mut.frame<-new.frame else mut.frame<-rbind(mut.frame,new.frame)
      #cat("mut.frame chr:",chr,"pos:",pos,"ref:",ref,"alt:",alt,"\n")
    } else {
      cat("\tSkipped\n")
    }
  }
  mut.frame<-mut.frame[order(mut.frame$transID),]  #because it takes time when transID differs from the previous one
  return(mut.frame)  
}

cDNAmut<-function(muttemp){ #build a matrix for mutation e.g. c.123A>G
  mut.frame<-data.frame()
  for (i in 1:length(muttemp)) {
    pos<-NA
    ref<-NA
    alt<-NA
    kind<-NA
    skipflag<-1  #In ver0.6 c.-14-51G>A c41+2T>A c53-11_53-7delCTT or ins are ignored
    temp<-regexpr("((\\-)*\\d+((\\+|\\-)\\d+)*)(_(\\-)*\\d+((\\+|\\-)\\d+)*)*",muttemp[i])
    pos<-substr(muttemp[i],as.numeric(temp),as.numeric(temp)+attr(temp,"match.length")-1)  #extract pos information like (-)123(+|-123)(_(-)345(+|-345))
    temp<-regexpr("(A|T|G|C)>(A|T|G|C)",muttemp[i])
    tempins<-regexpr("ins(A|T|G|C)+",muttemp[i])
    tempdel<-regexpr("del",muttemp[i])
    tempdelins<-regexpr("delins(A|T|G|C)+",muttemp[i])
    tempdup<-regexpr("dup",muttemp[i])
    tempdelins2<-regexpr("del(A|T|G|C)+(\\s+)*ins(A|T|G|C)+",muttemp[i])
    if (as.numeric(temp)!=-1) {   # G>A format
      ref<-substr(muttemp[i],as.numeric(temp),as.numeric(temp)) 
      alt<-substr(muttemp[i],as.numeric(temp)+2,as.numeric(temp)+2) 
      skipflag<-0
      kind<-"coding"
    } else if (as.numeric(tempdelins)!=-1 & attr(tempdelins,"match.length")>6) { #delinsATGC format
      ref<-"delins"
      alt<-substr(muttemp[i],as.numeric(tempdelins)+6,as.numeric(tempdelins)+attr(tempdelins,"match.length")-1)
      skipflag<-0
      kind<-"coding_delins"
    } else if (as.numeric(tempdelins2)!=-1 ) { #delATGCinsATGC format
      ref<-"del,ins"
      del<-NA
      ins<-NA
      temp<-regexpr("del(A|T|G|C)+",muttemp[i])  # take del seq
      if (as.numeric(temp)!=-1 & attr(temp,"match.length")>3 ) del<-substr(muttemp[i],as.numeric(temp)+3,as.numeric(temp)+attr(temp,"match.length")-1)
      temp<-regexpr("ins(A|T|G|C)+",muttemp[i])  #if dup seq given, take it
      if (as.numeric(temp)!=-1 & attr(temp,"match.length")>3 ) ins<-substr(muttemp[i],as.numeric(temp)+3,as.numeric(temp)+attr(temp,"match.length")-1)
      if (!is.na(del) & !is.na(ins)) {
        alt<-paste(del,ins,sep=",")
        skipflag<-0
        kind<-"coding_del_ins"
      }
    } else if (as.numeric(tempins)!=-1 & attr(tempins,"match.length")>3) {  #insATGC format
      ref<-"ins"  
      alt<-substr(muttemp[i],as.numeric(tempins)+3,as.numeric(tempins)+attr(tempins,"match.length")-1)
      skipflag<-0
      kind<-"coding_ins"
    } else if (as.numeric(tempdel)!=-1) {  #del(ATGC) format
      ref<-"del"
      alt<-""
      temp<-regexpr("del(A|T|G|C)+",muttemp[i])  #if del seq given, take it
      if (as.numeric(temp)!=-1 & attr(temp,"match.length")>3 ) alt<-substr(muttemp[i],as.numeric(temp)+3,as.numeric(temp)+attr(temp,"match.length")-1)
      skipflag<-0
      kind<-"coding_del"
    } else if (as.numeric(tempdup)!=-1) {  #dup(ATGC) format
      ref<-"dup"
      alt<-""
      temp<-regexpr("dup(A|T|G|C)+",muttemp[i])  #if dup seq given, take it
      if (as.numeric(temp)!=-1 & attr(temp,"match.length")>3 ) alt<-substr(muttemp[i],as.numeric(temp)+3,as.numeric(temp)+attr(temp,"match.length")-1)
      temp<-regexpr("dup\\d+",muttemp[i])  #if #dup given, take it
      if (as.numeric(temp)!=-1 & attr(temp,"match.length")>3 ) alt<-substr(muttemp[i],as.numeric(temp)+3,as.numeric(temp)+attr(temp,"match.length")-1)
      skipflag<-0
      kind<-"coding_dup"
    } 
    if (skipflag==0) {
      #      new.frame<-data.frame(kind=I("chrpos"),chr=I(chr),pos=as.integer(pos),ref=I(ref),alt=I(alt),name=I(paste("chr",chr,":",pos,ref,">",alt,sep="")),transID=I(transID))
      
      
      new.frame<-data.frame(kind=I(kind),chr=I(NA),pos=I(pos),ref=I(ref),alt=I(alt),name=I(paste("c.",pos,ref,ifelse(length(grep(">",muttemp[i]))>0,">",""),alt,sep="")),transID=I(IN$esbTranscriptID))
      if (nrow(mut.frame)==0) mut.frame<-new.frame else mut.frame<-rbind(mut.frame,new.frame)
    } 
  }
  return(mut.frame)  
}

cDNAposdetermine<-function(pos) {
  pos<-as.character(pos)
  if (length(grep("_",pos))>0) c<-as.character(unlist(strsplit(pos,"_"))) else c<-as.character(pos)
  if (length(c)>1) {
    if (length(grep("\\D",c[1]))==0 & length(grep("\\D",c[2]))==0) {
      c1temp<-as.numeric(c[1])
      c2temp<-as.numeric(c[2])
      if ((c1temp>c2temp) & (c2temp  == (c1temp %% 10^nchar(c[2]) +1 ) ) )  {  #if given c.123_4insAA then complement second position
        cat ("Digits could be omitted in second position of ",pos, ". Therefore it was complemented automatically.\n")
        c[2]<-as.character(c1temp+1) 
      }
    }
    newpos<-c(cDNApossub(c[1]),cDNApossub(c[2])) 
  }
  else newpos<-c(cDNApossub(c[1]),cDNApossub(c[1]))
  
  #cat("pos1:",newpos[1]," pos2:", newpos[2],"\n")
  return (newpos)
}

cDNApossub<-function(pos) {
  pos<-as.character(pos)
  minusflag<-0 #if you have c.-12, then 1
  jointflag<-0 #if you have 12+12 ,then 1 if you have 12-12 then 2   
  if (length(grep("^-",pos))>0) {minusflag<-1; pos <-sub("-","",pos);}
  if (length(grep("\\-",pos))>0) {c<-unlist(strsplit(pos,"\\-"));jointflag<-2;} else if (length(grep("\\+",pos))>0) {c<-unlist(strsplit(pos,"\\+"));jointflag<-1;} else c<-pos
  if (minusflag) {
    temp<-which(lseq$cDNApos==1)
    newpos<-temp-as.numeric(c[1])
  } else newpos<-which(lseq$cDNApos==as.numeric(c[1]))
  if (jointflag>0) {
    if (jointflag==1) newpos<-newpos+as.numeric(c[2]) else newpos<-newpos-as.numeric(c[2])
  }
  return(newpos)
}

filltheblank<-function (refseq,altseq,word="#",forward=FALSE){  #this function returns filled sentence with WORD if the length between refseq and altseq is different . e.g.  ref=AA  alt=AATTTG then  return AA####  (forward=FALSE)  ####AA (forward=TRUE)     
  returnword<-""
  diff<-0
  if (nchar(refseq) > nchar(altseq)) {
    returnword<-altseq
    diff<-nchar(refseq) - nchar(altseq)
  } else {
    returnword<-refseq
    diff<-nchar(altseq) - nchar(refseq)
  }
  addword<-""
  if (diff>0) for (i in 1:diff) {addword<-paste(addword,word,sep="");}
  returnword<- ifelse (forward,paste(addword, returnword,sep=""),paste(returnword,addword,sep="") )
  return(returnword)
}

reversenuc<-function(text) {  #reverse nucleotide and order of sequence   ex)input ATGC  -> output GCAT
  cseq<-c()
  returntext<-""
  for (i in 1:nchar(text)) {
    c<-substr(text,i,i)
    if (c=="t") newc<-"a"
    if (c=="a") newc<-"t"
    if (c=="c") newc<-"g"
    if (c=="g") newc<-"c"
    if (c==".") newc<-"."
    if (c=="T") newc<-"A"
    if (c=="A") newc<-"T"
    if (c=="C") newc<-"G"
    if (c=="G") newc<-"C"
    if (c==".") newc<-"."
    if (c=="(") newc<-")"
    if (c==")") newc<-"("
    if (c=="[") newc<-"]"
    if (c=="]") newc<-"["
    cseq<-c(cseq,newc)
  }
  for (i in nchar(text):1) {
    returntext<-paste(returntext,cseq[i],sep="")
  }
  return(returntext)
}

assignseq<-function(addstr,lseq,chr,startpos) {
  justbeforeintron<-0
  pos<-nrow(lseq)
  if (pos==0) {
    cDNApos<-1
    exon<-0
    justbeforeintron<-1
  } else {
    cDNApos<-max(lseq$cDNApos)+1; if (cDNApos==(-9+1)) cDNApos<-1
    exon<-max(lseq$exon); if(exon==-9) exon<-0
    if (lseq$cDNApos[pos]==-9) justbeforeintron<-1
  }
  pos<-pos+1
  phypos<-pos+startpos-1
  if (IN$seqinverse==1) phypos<-startpos+length(lseqtemp2)-pos
  if (length(grep("a|t|g|c|n",addstr))>0) { #intron part
    addstr<-toupper(addstr)
    new<-data.frame(ID="Givenseq",Seq=I(addstr),chr=chr,phypos=I(phypos),cDNApos=-9,exon=-9)
  } else if (length(grep("A|T|G|C|N",addstr))>0) { #exon part
    if (justbeforeintron) exon<-exon+1
    new<-data.frame(ID="Givenseq",Seq=I(addstr),chr=chr,phypos=I(phypos),cDNApos=cDNApos,exon=exon)
  }
  return(new)
}

minmaxfunc<-function(phypos) {
    min<-min(phypos)
    max<-max(phypos)
  if (reverseflag) {
    minmax<-list(min=max,max=min)
  } else {
    minmax<-list(min=min,max=max)
  }
 return (minmax)
}




#main
cat("This script is to calculate the integrated score of 5ss or 3ss\n")

#Input parametes -mutfile (chrpos format mut file with transcript ID) -output 
inputlist<-list(helpflag="-help",debug="-debug",marthost="-marthost",mutfile="-mutfile",output="-output",seq="-seq",seqinverse="-seqinverse",nomart="-nomart",
                esbTranscriptID="-esbTranscriptID",canonicalSS="-canonicalSS",entiregene="-entiregene",ss5="-ss5",ss3="-ss3",wide="-wide",lseq="-lseq",
                skipRegressScore="-skipRegressScore",skipSRE="-skipSRE"
                )
inputparas<-c(0,0,1,1,1,1,0,0,
              1,0,0,0,0,1,1,
              0,0
              )  #the length of parameters to be input in the command line 0-switch yes or no 1- one variable 2-two variables
IN<-inputparameters(inputlist,inputparas)
if (IN$helpflag==1) stop("\n",Helpdocument)
if (IN$output==-9) IN$output<-"Regress_Score.v0.5.R.result.txt"
if (IN$ss5==-9 & IN$ss3==-9) {IN$ss5<-1;IN$ss3<-1}

cat("Parameters:\n")
for (i in names(IN)) {
  cat(i,":",unlist(IN[which(names(IN)==i)]),"\n")
}

#initiation for using biomaRT
library(biomaRt) #if you need downloading biomaRt, please type as follows
#source("http://bioconductor.org/biocLite.R")
#biocLite("biomaRt")
#biocLite("BiocUpgrade")
#to use developer ver. download source file, then type install.packages("biomaRt_2.23.0.tar.gz", repos = NULL, type="source")

masterlseq<-data.frame()
if (IN$lseq!=-9) {
  IN$nomart<-1
  masterlseq<-read.table(IN$lseq,header=T,sep="\t",stringsAsFactors=F)
  if (IN$esbTranscriptID==-9) IN$esbTranscriptID<-IN$lseq
}
if (IN$nomart==-9) {
cat("Connecting biomaRt server...\n")
marthostname<-ifelse(IN$marthost==-9,"feb2014.archive.ensembl.org",IN$marthost) #feb2014.archive.ensembl.org   is hg19 latest ver host
cat("Host name is",marthostname,"\n")
mart = useMart("ENSEMBL_MART_ENSEMBL", dataset="hsapiens_gene_ensembl",host=marthostname)  # biomart for gene annotation
#host="may2009.archive.ensembl.org",  hg18
#host="Sep2013.archive.ensembl.org", hg19
} else {
  cat("Skip connecting biomaRt server.\n")
}

#preparing 3ss precalculated matrix
cat("Preparing maxentscan3ss parameters...\n")
metables<-makemaxentscores3()

#preparing 5ss precalculated matrix
cat("Preparing maxentscan5ss parameters...\n")
scorematrix<-makemaxentscores5()


#Read ESE and ESS consensus sequence
cat("Reading ESE and ESS sequences...\n")
ESEseq<-readLines(con="./ssfiles/ESE.txt")
ESElength<-nchar(ESEseq[1])

ESSseq<-readLines(con="./ssfiles/fas-hex3.txt")
ESSlength<-nchar(ESSseq[1])
cat("Reading ISE sequences...\n")
ISEseq<-readLines(con="./ssfiles/FAS-ISE.txt")
ISE9seq<-ISEseq[nchar(ISEseq)==9]
ISE9length<-9
ISE10seq<-ISEseq[nchar(ISEseq)==10]
ISE9length<-10

#set weight of each element and parameters in regression model
Eweight<-list(ESE5="47.84491,60.92733",ESE3="48.83209,60.70861",
              bGrun="44.53022,83.98231",aGrun="0.00000,58.65735",
              bISE="81.18222,14.99889",aISE="22,15,82,15",
              ESS5="42.17946,65.45010,203.29782,27.52051",ESS3="-63.3802067,15.25197,0.7984965,31.62696,86.4669410,59.74309,202.9456879,22.73068")
RegCoef5<-list(Intercept=-4.745344,MAXENT5=0.492312,ESE3=17.136712,ESS3=-20.178786,aGrun=80.972702,aISE=11.225050)
RegCoef3<-list(Intercept=-4.320291,MAXENT3=0.407358,ESE5=19.912353,ESS5=-31.136150,bGrun=61.247241,bISE=25.560157)

reverseflag<-0
muttemp<-c()
lseq<-data.frame()
newcDNApos<-c()
if (IN$seq!=-9) { #exon should be UPPERCASE, intron should be LOWERCASE, you can indicate mutation like (A|B), where A means refseq and B mean alt seq
                  #                                       Also you can indicate nucleutide of interest like ATGC[ATGC]ATGCa   
  cat("loading sequence information from:",IN$seq,"\n")
  data<-readLines(con=IN$seq)
  chr<-"-9"
  startpos<-1
  #clean up lines
  lseqtemp<-""
  if (length(data)>1) { 
    for (i in 1:length(data)) {
     if (length(grep(">",data[i]))) {
       temp<-regexpr("chr\\w+:\\d+",data[i]) #instead of variant position, actual 5ss gap taken
       if (as.numeric(temp)>0) {
          chrpos<-unlist(strsplit(substr(data[i],as.numeric(temp),as.numeric(temp)+attr(temp,"match.length")-1),":"))
          chr<-sub("chr","",chrpos[1])
          startpos<-as.numeric(chrpos[2])
       } 
     } else {
       lseqtemp<-paste(lseqtemp,data[i],sep="")
     }
    }
  } else {
   lseqtemp<-data[1]
  }
  ##clearnup the strings
  lseqtemp2<-c()
  if (IN$seqinverse==1) {start<-nchar(lseqtemp);end<-1} else {start<-1;end<-nchar(lseqtemp)} #if inverse switch on, then take characters inversely
  for (i in start:end) {  #construct string including necessary information ( and discard unnecessary characters)
    if (length(grep("a|t|g|c|n|A|T|G|C|N|(|)|[|]|\\|",substr(lseqtemp,i,i)))>0) {
      addstr<-substr(lseqtemp,i,i)
      if (IN$seqinverse==1) addstr<-reversenuc(addstr)
      lseqtemp2<-c(lseqtemp2,addstr)
    }
  }
  #construct sequence frame
  refposstart<-0
  refposend<-0
  concernstart<-0
  concernend<-0
  altseq<-""
  i<-1
  while (i<=length(lseqtemp2)) {
      addstr<-lseqtemp2[i]
      pos<-nrow(lseq)+1
      if (addstr=="(") {  #indicate mut like (ref|alt)
        refposstart<-pos
        i<-i+1
        tempchr<-lseqtemp2[i]
        loopflag<-1
        altcollect<-0
        while (loopflag) {
          if (tempchr==")") {
            loopflag<-0
          } else if (altcollect) {
            altseq<-paste(altseq,tempchr,sep="")
          } else if (tempchr=="|") {
            refposend<-nrow(lseq)
            altcollect<-1
          } else  {
            lseq<-rbind(lseq,assignseq(tempchr,lseq,chr,startpos))
          }
          if (loopflag) i<-i+1
          tempchr<-lseqtemp2[i]
        }
      } else if (addstr=="[") {
        concernstart<-pos
      } else if (addstr=="]") {
        concernend<-pos-1
      } else if (length(grep("a|t|g|c|n|A|T|G|C|N",addstr))>0) { 
        lseq<-rbind(lseq,assignseq(addstr,lseq,chr,startpos))
      }
      i<-i+1
  }
  if ((concernstart*concernend)!=0) {
    newcDNApos<-c(concernstart,concernend)
  } else if ((refposstart*refposend) !=0) {
    phypos<-refposstart+startpos-1
    if (IN$seqinverse==1) phypos<-startpos+length(lseqtemp2)-refposstart
    muttemp<-paste("chr",chr,":",phypos,paste(lseq$Seq[refposstart:refposend],collapse=""),">",altseq,sep="")
  }
}

if (IN$mutfile!=-9) {
  muttemp<-readLines(con=IN$mutfile)
}

mut.frame<-data.frame()
if (length(muttemp)>0) {
  if (length(grep("c\\.",muttemp))>0) mut.frame<-cDNAmut(muttemp)
  if (length(grep("chr",muttemp))>0) mut.frame<-chrposmut(muttemp) #new.frame<-data.frame(kind=I("chrpos"),chr=I(chr),pos=as.integer(pos),ref=I(ref),alt=I(alt),name=I(paste("chr",chr,":",pos,ref,">",alt,sep="")),transID=I(transID))
}

exon.frame<-data.frame(ID=I("DUMMY"),cDNApos=I("DUMMY"),exon=I("DUMMY")) #record original 5ss between exons
result.frame<-data.frame()

if (IN$debug==1) {
  write.table(lseq,file=paste("debug.lseq.",basename(IN$output),sep=""),col.names=T,row.names=F,quote=F)
  write.table(mut.frame,file=paste("debug.mut.table.",basename(IN$output),sep=""),col.names=T,row.names=F,quote=F)
}

#mut.frame loop
preTransID<-"dummy"
if (nrow(mut.frame)>0) { 
  for (i in 1:nrow(mut.frame)) {  #main loop for mut.frame
    cat("Processing",mut.frame$name[i],"   ",mut.frame$transID[i],"...\n")
  getseq<-0;if (!is.na(mut.frame$transID[i])) if (mut.frame$transID[i]!=preTransID) getseq<-1
  if (nrow(lseq)==0 | getseq) {lseq<-GetSeq(mut.frame$transID[i],1);preTransID<-mut.frame$transID[i]}
  if (length(grep("coding",mut.frame[i,]$kind))>0) newcDNApos<-cDNAposdetermine(mut.frame[i,]$pos) #which(lseq$cDNApos==mut.frame[i,]$pos)
if (mut.frame[i,]$kind=="chrpos" ) {newcDNApos<-which(lseq$phypos==mut.frame[i,]$pos);newcDNApos<-c(newcDNApos,newcDNApos);}
if (reverseflag==1 & mut.frame[i,]$kind=="chrpos") {
  cat("ChrPos information in a reverseStarnd gene. Therefore, change the allele info for cDNA.\n")
  mut.frame[i,]$ref<-reversenuc(mut.frame[i,]$ref)
  mut.frame[i,]$alt<-reversenuc(mut.frame[i,]$alt)
}
if (length(newcDNApos)>0) { #if cDNApos assigned correctly
newseq<-lseq
if (mut.frame[i,]$kind=="chrpos" & length(grep("^(A|T|G|C)$",mut.frame[i,]$ref))==1  & length(grep("^(A|T|G|C)$",mut.frame[i,]$alt))==1 | mut.frame[i,]$kind=="coding" ) {
  newseq[newcDNApos[1],]$Seq<-mut.frame[i,]$alt;indelflag<-0;
} else if (mut.frame[i,]$kind=="coding_ins") {
  insert<-data.frame()
  for (j in 1:nchar(mut.frame[i,]$alt)) {
    temp<-lseq[1,]
    temp$Seq<-substr(mut.frame[i,]$alt,j,j)
    temp$phypos<-paste(lseq[newcDNApos[1],]$phypos,"+ins",j,sep="")
    temp$cDNApos<-paste(lseq[newcDNApos[1],]$cDNApos,"+ins",j,sep="")
    temp$exon<-lseq[newcDNApos[1],]$exon
    insert<-rbind(insert,temp)
  }
  newseq<-rbind(newseq[1:newcDNApos[1],],insert,newseq[newcDNApos[2]:nrow(newseq),])
  newcDNApos[1]<-newcDNApos[1]+1
  newcDNApos[2]<-newcDNApos[2]+nrow(insert)-1
  
} else if (mut.frame[i,]$kind=="coding_del") {
  delseq<-""
  for (j in newcDNApos[1]:newcDNApos[2]) {delseq<-paste(delseq,newseq[j,]$Seq,sep="");}
  #cat("Deleted Seq is :",toupper(delseq),"mut.alt:",mut.frame[i,]$alt,"\n")####
  if (nchar(mut.frame[i,]$alt)==0 | is.na(mut.frame[i,]$alt)) cat("Deleted Seq is :",toupper(delseq),"\n") else if (toupper(delseq)!=toupper(mut.frame[i,]$alt)) cat("Warning. Given del seq:",toupper(mut.frame[i,]$alt),"is different from actual cDNA seq:",toupper(delseq),"\n")
  newseq<-rbind(newseq[1:(newcDNApos[1]-1),],newseq[(newcDNApos[2]+1):nrow(newseq),])
  #newcDNApos[1]<-newcDNApos[1]
  newcDNApos[2]<-newcDNApos[1]-1
} else if (mut.frame[i,]$kind=="coding_delins") {
  insert<-data.frame()
  for (j in 1:nchar(mut.frame[i,]$alt)) {
    temp<-lseq[1,]
    temp$Seq<-substr(mut.frame[i,]$alt,j,j)
    temp$phypos<-paste(lseq[newcDNApos[1],]$phypos,"+ins",i,sep="")
    temp$cDNApos<-paste(lseq[newcDNApos[1],]$cDNApos,"+ins",j,sep="")
    temp$exon<-lseq[newcDNApos[1],]$exon
    insert<-rbind(insert,temp)
  }
  newseq<-rbind(newseq[1:(newcDNApos[1]-1),],insert,newseq[(newcDNApos[2]+1):nrow(newseq),])
  newcDNApos[2]<-newcDNApos[1]+nrow(insert)-1
} else if (mut.frame[i,]$kind=="coding_del_ins") {
  delins<-unlist(strsplit(mut.frame[i,]$alt,","))
  delseq<-"";
  for (j in newcDNApos[1]:newcDNApos[2]) {delseq<-paste(delseq,newseq[j,]$Seq,sep="");}
  if (toupper(delseq)!=toupper(delins[1])) cat("Warning. Given del seq:",toupper(mut.frame[i,]$alt),"is different from actual cDNA seq:",toupper(delseq),"\n")
  insert<-data.frame()
  for (j in 1:nchar(mut.frame[i,]$alt)) {
    temp<-lseq[1,]
    temp$Seq<-substr(mut.frame[i,]$alt,j,j)
    temp$phypos<-paste(lseq[newcDNApos[1],]$phypos,"+ins",j,sep="")
    temp$cDNApos<-paste(lseq[newcDNApos[1],]$cDNApos,"+ins",j,sep="")
    temp$exon<-lseq[newcDNApos[1],]$exon
    insert<-rbind(insert,temp)
  }
  newseq<-rbind(newseq[1:(newcDNApos[1]-1),],insert,newseq[(newcDNApos[2]+1):nrow(newseq),])
  newcDNApos[2]<-newcDNApos[1]+nrow(insert)-1
} else if (mut.frame[i,]$kind=="coding_dup") {
  insertseq<-"";
  for (j in newcDNApos[1]:newcDNApos[2]) {
    insertseq<-paste(insertseq,toupper(lseq[j,]$Seq),sep="")
  }
  if (length(grep("\\d+",mut.frame[i,]$alt))>0) {
    insertseq2<-"";
    for (j in 1:(as.numeric(mut.frame[i,]$alt)-1)) {
      insertseq2<-paste(insertseq2,insertseq,sep="")
    }
  } else if (nchar(mut.frame[i,]$alt)==0) {
    insertseq2<-insertseq
  } else {
    insertseq2<-toupper(mut.frame[i,]$alt)
    #cat ("insertseq",insertseq,"insertseq2",insertseq2,"i",i,"ref",mut.frame[i,]$ref,"alt",class(mut.frame[i,]$alt),"\n")
    #warnings()
    if (insertseq!=substr(insertseq2,1,nchar(insertseq))) cat ("Warning. Given dup seq:",insertseq2,"is different from actual cDNA seq:",insertseq,"\n")
  }
  insert<-data.frame()
  for (j in 1:nchar(insertseq2)) {
    temp<-lseq[1,]
    temp$Seq<-substr(insertseq2,j,j)
    temp$phypos<-paste(lseq[newcDNApos[1],]$phypos,"+dup",j,sep="")
    temp$cDNApos<-paste(lseq[newcDNApos[2],]$cDNApos,"+dup",j,sep="")
    temp$exon<-lseq[newcDNApos[2],]$exon
    insert<-rbind(insert,temp)
  }
  newseq<-rbind(newseq[1:(newcDNApos[2]),],insert,newseq[(newcDNApos[2]+1):nrow(newseq),])
  newcDNApos[1]<-newcDNApos[2]+1
  newcDNApos[2]<-newcDNApos[2]+nrow(insert)
} else if (mut.frame[i,]$kind=="chrpos") { #in case of indel of vcf file
  delseq<-"";
  ncommon<-0;
  
  if (mut.frame[i,]$ref=="." ) mut.frame[i,]$ref<-""
  if (mut.frame[i,]$alt=="." ) mut.frame[i,]$alt<-""
  ref <-mut.frame[i,]$ref
  alt <-mut.frame[i,]$alt
  add<-ifelse(nchar(mut.frame[i,]$alt)>0,1,0)  #for length(alt)==0 
  if (reverseflag==1) {
    temprefalt<-filltheblank(ref,alt,word="#",forward=TRUE)
    if (nchar(ref)<nchar(alt)) ref<-temprefalt else if (nchar(ref)>nchar(alt)) alt<-temprefalt
    conflag<-1;
    newref<-""  # new ref seq where common charcters omitted
    newalt<-"" # new alt seq where common characters omitted
    for (j in nchar(ref):1) {
      if ( substr(ref,j,j)==substr(alt,j,j) & conflag==1 ) {ncommon<-ncommon+1} else {conflag<-0;newref<-paste(ifelse(substr(ref,j,j)!="#",substr(ref,j,j),""),newref,sep="");newalt<-paste(ifelse(substr(alt,j,j)!="#",substr(alt,j,j),""),newalt,sep="")}
    }
    cat("newref:",newref," newalt:",newalt,"\n")
    refstart<-newcDNApos[1]-nchar(mut.frame[i,]$ref) #just left position of substitution
    refend<-newcDNApos[1]-ncommon+1 #just right position of substitution
  } else {
    temprefalt<-filltheblank(ref,alt,word="#",forward=FALSE)
    if (nchar(ref)<nchar(alt)) ref<-temprefalt else if (nchar(ref)>nchar(alt)) alt<-temprefalt
    conflag<-1;
    newref<-""  # new ref seq where common charcters omitted
    newalt<-"" # new alt seq where common characters omitted
    for (j in 1:nchar(ref)) {
      if ( substr(ref,j,j)==substr(alt,j,j) & conflag==1 ) {ncommon<-ncommon+1} else {conflag<-0;newref<-paste(newref,ifelse(substr(ref,j,j)!="#",substr(ref,j,j),""),sep="");newalt<-paste(newalt,ifelse(substr(alt,j,j)!="#",substr(alt,j,j),""),sep="")}
    }
    cat("newref:",newref," newalt:",newalt,"\n")
    refstart<-newcDNApos[1]+ncommon-1 #just left position of substitution
    refend<-newcDNApos[1]+nchar(mut.frame[i,]$ref) #just right position of substitution
  }
  newcDNApos[1]<-refstart+1
  newcDNApos[2]<-refstart+nchar(newalt)
  
  for (j in (refstart+1-ifelse(reverseflag==0,ncommon,0)):(refend-1+ifelse(reverseflag==1,ncommon,0))) {delseq<-paste(delseq,newseq[j,]$Seq,sep="");}
  if (length(grep("^(A|T|G|C)+$",mut.frame[i,]$ref))==1 & mut.frame[i,]$ref == delseq | nchar(mut.frame[i,]$ref)==0) {# at 1st, check if ref seq is consistent with given ref information
    cat("Refseq:",mut.frame[i,]$ref ,"  Altseq:",mut.frame[i,]$alt,"\n")
    insert<-data.frame() #make insert seq  from alt seq
    subcount<-1 #substitution count
    insseq<-""
    if (nchar(newalt)>0) for (j in 1:nchar(newalt)) {
      temp<-lseq[1,]
      temp$Seq<-substr(newalt,j,j)
      temp$phypos<-paste(lseq[refstart,]$phypos,ifelse(reverseflag==1,"-","+"),"ins",j,sep="")
      tempcDNApos<-ifelse(lseq[refstart,]$cDNApos==-9,-9,lseq[refstart,]$cDNApos)
      temp$cDNApos<-paste(tempcDNApos,"+ins",j,sep="")
      temp$exon<-lseq[refstart,]$exon
      insert<-rbind(insert,temp)
    }
    newseq<-rbind(newseq[1:refstart,],insert,newseq[refend:nrow(newseq),])
    
  } else {
    cat("Warning. Given ref seq in vcf:",toupper(mut.frame[i,]$ref),"is different from actual cDNA seq:",toupper(delseq),"\nAnalysis continued, though.\n")
  }
}

#lseq<-data.frame(GENE=rep(transcriptinfo$hgnc_symbol,lseqlength),ID=rep(transciptID,lseqlength),chr=rep(transcriptinfo$chromosome_name,lseqlength),phypos=transcriptinfo$transcript_start:transcriptinfo$transcript_end,cDNApos=rep(-9,lseqlength),exon=rep(-9,lseqlength),Seq=I(unlist(strsplit(lseqtemp$transcript_exon_intron,""))))
#SSscan<-function(seq,ExPosInTheContext,SS)  {#seq a sequence including 5ss or 3ss, ExPosInTheContext exon position(Start,End) in the given seq  SS should be 5(5Splice site) or 3 (3Spice site) 
if (IN$debug==1) write.table(newseq,file=paste("debug.newseq.",basename(IN$output),sep=""),col.names=T,row.names=F,quote=F)

if (IN$wide!=-9) {
  cat("Scan region ", IN$wide, "bp attached around the indicated mutation area.\n")
  newcDNApos[1]<-newcDNApos[1]-as.numeric(IN$wide)
  newcDNApos[2]<-newcDNApos[2]+as.numeric(IN$wide)
  if (newcDNApos[1]<1) newcDNApos[1]<-1
  if (newcDNApos[2]>nrow(newseq)) newcDNApos[2]<-nrow(newseq)
}

#scan for 5ss
result5<-data.frame()
if (IN$ss5==1) {
  cat("Checking the potential of 5\'splice site...\n")
  Start<-newcDNApos[1]-9+1
  if (Start<1) Start<-1  #indicate the region where the mutation affects
  End <-newcDNApos[2]
  if (End > (nrow(newseq)-9+1)) End<-nrow(newseq)-9+1
  #cat("SS5 search pos1:",newcDNApos[1], " pos2:",newcDNApos[2], "Start:",Start," End", End, "\n")
  result5<-SS5scan(newseq,Start,End,newcDNApos)
  if (IN$debug==1) write.table(result5,file=paste("debug.result5.",basename(IN$output),sep=""),col.names=T,row.names=F,quote=F)
}


#scan for 3ss
result3<-data.frame()
if (IN$ss3==1) {
  cat("Checking the potential of 3\'splice site...\n")
  Start<-newcDNApos[1]-23+1
  if (Start<1) Start<-1  #indicate the region where the mutation affects
  End <-newcDNApos[2]
  if (End > (nrow(newseq)-23+1)) end<-nrow(newseq)-23+1
  #cat("SS3 search pos1:",newcDNApos[1], " pos2:",newcDNApos[2], "Start:",Start," End", End, "\n")
  result3<-SS3scan(newseq,Start,End,newcDNApos)
  if (IN$debug==1) write.table(result3,file=paste("debug.result3.",basename(IN$output),sep=""),col.names=T,row.names=F,quote=F)
}



if (IN$ss3==-9) {
  minmax<-minmaxfunc(result5$phypos)
} else if (IN$ss5==-9) {
  minmax<-minmaxfunc(result3$phypos)
} else {
  minmax<-minmaxfunc(c(result3$phypos,result5$phypos))
}
min<-minmax$min
max<-minmax$max

#-entiregene
if (IN$entiregene==1) {
  cat("Checking entire sequence of the gene. Then it could take time...\n")
  min<-lseq$phypos[1]
  max<-lseq$phypos[nrow(lseq)]
}

#calc scores in ref seq
cat("Calculating Scores in the referrence sequence...\n")

lseqmin<-which(lseq$phypos==min)
lseqmax<-which(lseq$phypos==max)
if (IN$ss5==1) {
  result5ref<-SS5scan(lseq,lseqmin,lseqmax,newcDNApos)
  if (IN$debug==1) write.table(result5ref,file=paste("debug.result5ref.",basename(IN$output),sep=""),col.names=T,row.names=F,quote=F)
}
if (IN$ss3==1) {
  result3ref<-SS3scan(lseq,lseqmin,lseqmax,newcDNApos)
  if (IN$debug==1) write.table(result3ref,file=paste("debug.result3ref.",basename(IN$output),sep=""),col.names=T,row.names=F,quote=F)
}

#integrate result
cat("Summarizing the results...\n")

for (j in min:max) {
  if (IN$ss5==1) {
  ss5pos<-which(result5ref$phypos==j)
  if (length(ss5pos)>0) {
    ss5lineref<-result5ref[ss5pos,-1]  #remove nrow column
  } else {
    ss5lineref<-data.frame(MaxEntScoreSS5=-999,SS5score=-999,SS5cDNAboundary="-",SS5phyposboundary="-",SS5seq="-",ESE5=-999,ESS5=-999,Grun5=-999,ISE5=-999)
  }
  names(ss5lineref)<-paste("Ref_",names(ss5lineref),sep="")
  ss5pos<-which(result5$phypos==j)
  if (length(ss5pos)>0) {
    ss5line<-result5[ss5pos,-1]  #remove nrow column
  } else {
    ss5line<-data.frame(MaxEntScoreSS5=-999,SS5score=-999,SS5cDNAboundary="-",SS5phyposboundary="-",SS5seq="-",ESE5=-999,ESS5=-999,Grun5=-999,ISE5=-999)
  }
  names(ss5line)<-paste("Alt_",names(ss5line),sep="")
  if (ss5lineref$Ref_MaxEntScoreSS5!=-999 & ss5line$Alt_MaxEntScoreSS5!=-999 ) diff5<-data.frame(Diff_MaxEntScoreSS5=(ss5line$Alt_MaxEntScoreSS5-ss5lineref$Ref_MaxEntScoreSS5),Diff_SS5score=(ss5line$Alt_SS5score-ss5lineref$Ref_SS5score),
                                                                                                 Diff_ESE5=(ss5line$Alt_ESE5-ss5lineref$Ref_ESE5),Diff_ESS5=(ss5line$Alt_ESS5-ss5lineref$Ref_ESS5),Diff_Grun5=(ss5line$Alt_Grun5-ss5lineref$Ref_Grun5),Diff_ISE5=(ss5line$Alt_ISE5-ss5lineref$Ref_ISE5))
  else diff5<-data.frame(Diff_MaxEntScoreSS5=-999,Diff_SS5score=-999,
                         Diff_ESE5=-999,Diff_ESS5=-999,Diff_Grun5=-999,Diff_ISE5=-999)
  }

  if (IN$ss3==1) {
  ss3pos<-which(result3ref$phypos==j)
  if (length(ss3pos)>0) {
    ss3lineref<-result3ref[ss3pos,-1]#remove nrow column
  } else {
    ss3lineref<-data.frame(MaxEntScoreSS3=-999,SS3score=-999,SS3cDNAboundary="-",SS3phyposboundary="-",SS3seq="-",ESE3=-999,ESS3=-999,Grun3=-999,ISE3=-999)
  }
  names(ss3lineref)<-paste("Ref_",names(ss3lineref),sep="")
  ss3pos<-which(result3$phypos==j)
  if (length(ss3pos)>0) {
    ss3line<-result3[ss3pos,-1]#remove nrow column
  } else {
    ss3line<-data.frame(MaxEntScoreSS3=-999,SS3score=-999,SS3cDNAboundary="-",SS3phyposboundary="-",SS3seq="-",ESE3=-999,ESS3=-999,Grun3=-999,ISE3=-999)
  }
  names(ss3line)<-paste("Alt_",names(ss3line),sep="")
  #cat("#row newseq:",nrow(newseq[j,]), " #row ss5line:",nrow(ss5line),"#row ss3line:",nrow(ss3line),"\n")
  if (ss3lineref$Ref_MaxEntScoreSS3!=-999 & ss3line$Alt_MaxEntScoreSS3!=-999 ) diff3<-data.frame(Diff_MaxEntScoreSS3=(ss3line$Alt_MaxEntScoreSS3-ss3lineref$Ref_MaxEntScoreSS3),Diff_SS3score=(ss3line$Alt_SS3score-ss3lineref$Ref_SS3score),
                                                                                                 Diff_ESE3=(ss3line$Alt_ESE3-ss3lineref$Ref_ESE3),Diff_ESS3=(ss3line$Alt_ESS3-ss3lineref$Ref_ESS3),Diff_Grun3=(ss3line$Alt_Grun3-ss3lineref$Ref_Grun3),Diff_ISE3=(ss3line$Alt_ISE3-ss3lineref$Ref_ISE3))
  else diff3<-data.frame(Diff_MaxEntScoreSS3=-999,Diff_SS3score=-999,
                         Diff_ESE3=-999,Diff_ESS3=-999,Diff_Grun3=-999,Diff_ISE3=-999)
  }
  
  if (IN$ss5==1 & IN$ss3==1) {
    new<-cbind(data.frame(Var_name=mut.frame$name[i]),newseq[which(newseq$phypos==j),],ss5lineref,ss5line,diff5,ss3lineref,ss3line,diff3)
  } else if (IN$ss5==1 & IN$ss3==-9) {
    new<-cbind(data.frame(Var_name=mut.frame$name[i]),newseq[which(newseq$phypos==j),],ss5lineref,ss5line,diff5)
  } else {
    new<-cbind(data.frame(Var_name=mut.frame$name[i]),newseq[which(newseq$phypos==j),],ss3lineref,ss3line,diff3)
  }
  result.frame<-rbind(result.frame,new)
}

#if -canonicalSS on
if (IN$canonicalSS==1 & newseq$exon[newcDNApos[1]]!=-9 & newseq$exon[newcDNApos[2]]!=-9) {  #the variant should be on an exon
  cat("Calculating Score in the canonical splice sites nearby...\n")
  #searchin for 5 prime splice site
  nexon<-newseq$exon[newcDNApos[1]]
  if (IN$ss5==1) {
  pos5ss<-0
  for (j in newcDNApos[2]:nrow(newseq)) {
    if (newseq$exon[j]==-9) {pos5ss<-j-1; break}
  }
  if (pos5ss>2) {
    result5ss_in_alt<-SS5scan(newseq,pos5ss-2,pos5ss-2,c(-9,-9))
    refpos<-which(lseq$phypos==newseq$phypos[pos5ss])
    result5ss_in_ref<-SS5scan(lseq,refpos-2,refpos-2,c(-9,-9))
    ss5lineref<-result5ss_in_ref[,-1]
    ss5line<-result5ss_in_alt[,-1]
    names(ss5lineref)<-paste("Ref_",names(ss5lineref),sep="")
    names(ss5line)<-paste("Alt_",names(ss5line),sep="")
    if (ss5lineref$Ref_MaxEntScoreSS5!=-999 & ss5line$Alt_MaxEntScoreSS5!=-999 ) diff5<-data.frame(Diff_MaxEntScoreSS5=(ss5line$Alt_MaxEntScoreSS5-ss5lineref$Ref_MaxEntScoreSS5),Diff_SS5score=(ss5line$Alt_SS5score-ss5lineref$Ref_SS5score),
                                                                                                   Diff_ESE5=(ss5line$Alt_ESE5-ss5lineref$Ref_ESE5),Diff_ESS5=(ss5line$Alt_ESS5-ss5lineref$Ref_ESS5),Diff_Grun5=(ss5line$Alt_Grun5-ss5lineref$Ref_Grun5),Diff_ISE5=(ss5line$Alt_ISE5-ss5lineref$Ref_ISE5))
    else diff5<-data.frame(Diff_MaxEntScoreSS5=-999,Diff_SS5score=-999,
                           Diff_ESE5=-999,Diff_ESS5=-999,Diff_Grun5=-999,Diff_ISE5=-999)
    if (IN$ss3==1) {
      new<-cbind(data.frame(Var_name=mut.frame$name[i]),newseq[pos5ss-2,],ss5lineref,ss5line,diff5,
               data.frame(Ref_MaxEntScoreSS3=-999,Ref_SS3score=-999,Ref_SS3cDNAboundary="-",Ref_SS3phyposboundary="-",Ref_SS3seq="-",Alt_MaxEntScoreSS3=-999,Alt_SS3score=-999,Alt_SS3cDNAboundary="-",Alt_SS3phyposboundary="-",Alt_SS3seq="-",Diff_MaxEntScoreSS3=-999,Diff_SS3score=-999,Diff_ESE3=-999,Diff_ESS3=-999,Diff_Grun3=-999,Diff_ISE3=-999))
    } else {
      new<-cbind(data.frame(Var_name=mut.frame$name[i]),newseq[pos5ss-2,],ss5lineref,ss5line,diff5)
    }
      result.frame<-rbind(result.frame,new)
  }
  }
  
  if (IN$ss3==1) {
  pos3ss<-0
  for (j in newcDNApos[1]:1) {
    if (newseq$exon[j]==-9) {pos3ss<-j+1; break}
  }
  if (pos3ss>20)  {
    result3ss_in_alt<-SS3scan(newseq,pos3ss-20,pos3ss-20,c(-9,-9))
    refpos<-which(lseq$phypos==newseq$phypos[pos3ss])
    result3ss_in_ref<-SS3scan(lseq,refpos-20,refpos-20,c(-9,-9))
    ss3lineref<-result3ss_in_ref[,-1]
    ss3line<-result3ss_in_alt[,-1]
    names(ss3lineref)<-paste("Ref_",names(ss3lineref),sep="")
    names(ss3line)<-paste("Alt_",names(ss3line),sep="")
    if (ss3lineref$Ref_MaxEntScoreSS3!=-999 & ss3line$Alt_MaxEntScoreSS3!=-999 ) diff3<-data.frame(Diff_MaxEntScoreSS3=(ss3line$Alt_MaxEntScoreSS3-ss3lineref$Ref_MaxEntScoreSS3),Diff_SS3score=(ss3line$Alt_SS3score-ss3lineref$Ref_SS3score),
                                                                                                   Diff_ESE3=(ss3line$Alt_ESE3-ss3lineref$Ref_ESE3),Diff_ESS3=(ss3line$Alt_ESS3-ss3lineref$Ref_ESS3),Diff_Grun3=(ss3line$Alt_Grun3-ss3lineref$Ref_Grun3),Diff_ISE3=(ss3line$Alt_ISE3-ss3lineref$Ref_ISE3))
    else diff3<-data.frame(Diff_MaxEntScoreSS3=-999,Diff_SS3score=-999,Diff_ESE3=-999,Diff_ESS3=-999,Diff_Grun3=-999,Diff_ISE3=-999)
    if (IN$ss5==1) {
    new<-cbind(data.frame(Var_name=mut.frame$name[i]),newseq[pos3ss-20,],data.frame(Ref_MaxEntScoreSS5=-999,Ref_SS5score=-999,Ref_SS5cDNAboundary="-",Ref_SS5phyposboundary="-",Ref_SS5seq="-",Alt_MaxEntScoreSS5=-999,Alt_SS5score=-999,Alt_SS5cDNAboundary="-",Alt_SS5phyposboundary="-",Alt_SS5seq="-",Diff_MaxEntScoreSS5=-999,Diff_SS5score=-999,Diff_ESE5=-999,Diff_ESS5=-999,Diff_Grun5=-999,Diff_ISE5=-999),
               ss3lineref,ss3line,diff3)
    } else {
      new<-cbind(data.frame(Var_name=mut.frame$name[i]),newseq[pos3ss-20,],ss3lineref,ss3line,diff3)
    }
    result.frame<-rbind(result.frame,new)
  }
  }
}
} else {  #if cDNApos not assigned
  cat("Variant:",mut.frame$name[i], " is NOT on the indicated transcript:",mut.frame$transID[i], "propably because it is on the downstream or UTR regions.\n")
}

}  #mut.frame BIG i loop end
} else {  #just treat ref seq
  cat("Calculating Scores just in the referrence sequence...\n")
  if (IN$ss5==1) {
  cat("Checking the potential of 5\'splice site...\n")
  Start<-newcDNApos[1]-9+1
  if (Start<1) Start<-1  #indicate the region where the mutation affects
  End <-newcDNApos[2]
  if (End > (nrow(lseq)-9+1)) End<-nrow(lseq)-9+1
  result5ref<-SS5scan(lseq,Start,End,newcDNApos)
  }
  if (IN$ss3==1) {
  cat("Checking the potential of 3\'splice site...\n")
  Start<-newcDNApos[1]-23+1
  if (Start<1) start<-1  #indicate the region where the mutation affects
  End <-newcDNApos[2]
  if (End > (nrow(lseq)-23+1)) end<-nrow(lseq)-23+1
  result3ref<-SS3scan(lseq,Start,End,newcDNApos)
  }
  
  #integrate result
  cat("Summarizing the results...\n")
  if (IN$ss3==-9) {
    minmax<-minmaxfunc(result5ref$phypos)
  } else if (IN$ss5==-9) {
    minmax<-minmaxfunc(result3ref$phypos)
  } else {
    minmax<-minmaxfunc(c(result3ref$phypos,result5ref$phypos))
  }
  min<-minmax$min
  max<-minmax$max
  
  for (j in min:max) {
    if (IN$ss5==1) {
    ss5pos<-which(result5ref$phypos==j)
    if (length(ss5pos)>0) {
      ss5lineref<-result5ref[ss5pos,-1]  #remove nrow column
    } else {
      ss5lineref<-data.frame(MaxEntScoreSS5=-999,SS5score=-999,SS5cDNAboundary="-",SS5phyposboundary="-",SS5seq="-",ESE5="-",ESS5="-",Grun5="-",ISE5="-")
    }
    names(ss5lineref)<-paste("Ref_",names(ss5lineref),sep="")
    }
    if (IN$ss3==1) {
    ss3pos<-which(result3ref$phypos==j)
    if (length(ss3pos)>0) {
      ss3lineref<-result3ref[ss3pos,-1]#remove nrow column
    } else {
      ss3lineref<-data.frame(MaxEntScoreSS3=-999,SS3score=-999,SS3cDNAboundary="-",SS3phyposboundary="-",SS3seq="-",,ESE3="-",ESS3="-",Grun3="-",ISE3="-")
    }
    names(ss3lineref)<-paste("Ref_",names(ss3lineref),sep="")
    }
    if (IN$ss5==1 & IN$ss3==1) {
      new<-cbind(lseq[which(lseq$phypos==j),],ss5lineref,ss3lineref)
    } else if (IN$ss5==1 & IN$ss3==-9) {
      new<-cbind(lseq[which(lseq$phypos==j),],ss5lineref)
    } else {
      new<-cbind(lseq[which(lseq$phypos==j),],ss3lineref)
    }
    
    result.frame<-rbind(result.frame,new)
  }
}


#save result
cat("Writing splice site estimation result to:",IN$output,"\n")
write.table(result.frame,file=IN$output,row.names=F,col.names=T,sep="\t",quote=F)
warnings()











